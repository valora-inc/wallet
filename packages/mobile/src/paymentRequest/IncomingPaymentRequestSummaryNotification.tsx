import * as React from 'react'
import { WithTranslation } from 'react-i18next'
import { Image } from 'react-native'
import { connect } from 'react-redux'
import { HomeEvents } from 'src/analytics/Events'
import ValoraAnalytics from 'src/analytics/ValoraAnalytics'
import { NotificationBannerCTATypes, NotificationBannerTypes } from 'src/home/NotificationBox'
import { withTranslation } from 'src/i18n'
import { notificationIncomingRequest } from 'src/images/Images'
import { navigate } from 'src/navigator/NavigationService'
import { Screens } from 'src/navigator/Screens'
import SummaryNotification from 'src/notifications/SummaryNotification'
import { listItemRenderer } from 'src/paymentRequest/IncomingPaymentRequestListScreen'
import PaymentRequestNotificationInner from 'src/paymentRequest/PaymentRequestNotificationInner'
import { PaymentRequest } from 'src/paymentRequest/types'
import { getRecipientFromAddress, RecipientInfo } from 'src/recipients/recipient'
import { recipientInfoSelector } from 'src/recipients/reducer'
import { RootState } from 'src/redux/reducers'

interface OwnProps {
  requests: PaymentRequest[]
}

type Props = OwnProps & WithTranslation & StateProps

interface StateProps {
  recipientInfo: RecipientInfo
}

const mapStateToProps = (state: RootState): StateProps => ({
  recipientInfo: recipientInfoSelector(state),
})

// Payment Request notification for the notification center on home screen
export class IncomingPaymentRequestSummaryNotification extends React.Component<Props> {
  onReview = () => {
    ValoraAnalytics.track(HomeEvents.notification_select, {
      notificationType: NotificationBannerTypes.incoming_tx_request,
      selectedAction: NotificationBannerCTATypes.review,
    })
    navigate(Screens.IncomingPaymentRequestListScreen)
  }

  itemRenderer = (item: PaymentRequest) => {
    return (
      <PaymentRequestNotificationInner
        key={item.uid}
        amount={item.amount}
        recipient={getRecipientFromAddress(item.requesterAddress, this.props.recipientInfo)}
        t={this.props.t}
      />
    )
  }

  render() {
    const { recipientInfo, requests, t } = this.props

    return requests.length === 1 ? (
      listItemRenderer({
        recipientInfo,
      })(requests[0])
    ) : (
      <SummaryNotification<PaymentRequest>
        items={requests}
        title={t('incomingPaymentRequestsSummaryTitle', { count: requests.length })}
        detailsI18nKey="incomingPaymentRequestsSummaryDetails"
        icon={<Image source={notificationIncomingRequest} resizeMode="contain" />}
        onReview={this.onReview}
        itemRenderer={this.itemRenderer}
      />
    )
  }
}

export default connect<StateProps, {}, {}, RootState>(
  mapStateToProps,
  {}
)(withTranslation<Props>()(IncomingPaymentRequestSummaryNotification))
