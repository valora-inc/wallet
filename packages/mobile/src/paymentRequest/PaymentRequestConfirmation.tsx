import { firebase } from '@react-native-firebase/database'
import { StackScreenProps } from '@react-navigation/stack'
import React, { useState } from 'react'
import { useTranslation } from 'react-i18next'
import { StyleSheet, Text, View } from 'react-native'
import { SafeAreaView } from 'react-native-safe-area-context'
import { useDispatch } from 'react-redux'
import { e164NumberSelector } from 'src/account/selectors'
import { RequestEvents } from 'src/analytics/Events'
import ValoraAnalytics from 'src/analytics/ValoraAnalytics'
import BackButton from 'src/components/BackButton'
import CommentTextInput from 'src/components/CommentTextInput'
import ContactCircle from 'src/components/ContactCircle'
import ReviewFrame from 'src/components/ReviewFrame'
import TokenDisplay from 'src/components/TokenDisplay'
import TokenTotalLineItem from 'src/components/TokenTotalLineItem'
import { emptyHeader } from 'src/navigator/Headers'
import { Screens } from 'src/navigator/Screens'
import { StackParamList } from 'src/navigator/types'
import { writePaymentRequest } from 'src/paymentRequest/actions'
import { PaymentRequestStatus } from 'src/paymentRequest/types'
import { getDisplayName } from 'src/recipients/recipient'
import useSelector from 'src/redux/useSelector'
import { useInputAmounts } from 'src/send/SendAmount'
import { useRecipientToSendTo } from 'src/send/SendConfirmation'
import DisconnectBanner from 'src/shared/DisconnectBanner'
import colors from 'src/styles/colors'
import fontStyles from 'src/styles/fonts'
import Logger from 'src/utils/Logger'
import { walletAddressSelector } from 'src/web3/selectors'

// @ts-ignore
const TAG = 'paymentRequest/confirmation'

type Props = StackScreenProps<StackParamList, Screens.PaymentRequestConfirmation>

export const paymentConfirmationScreenNavOptions = () => ({
  ...emptyHeader,
  headerLeft: () => <BackButton eventName={RequestEvents.request_confirm_back} />,
})

function PaymentRequestConfirmation({ route }: Props) {
  const [comment, setComment] = useState('')
  const { transactionData } = route.params
  const requesterAddress = useSelector(walletAddressSelector)
  const requesterE164Number = useSelector(e164NumberSelector)

  const recipient = useRecipientToSendTo(transactionData.recipient)
  const { tokenAmount, usdAmount } = useInputAmounts(
    transactionData.inputAmount.toString(),
    transactionData.amountIsInLocalCurrency,
    transactionData.tokenAddress
  )

  const { t } = useTranslation()

  const dispatch = useDispatch()

  const onBlur = () => {
    const trimmedComment = comment.trim()
    setComment(trimmedComment)
  }

  const onConfirm = async () => {
    if (!recipient) {
      throw new Error("Can't request without valid recipient")
    }

    if (!requesterAddress) {
      throw new Error("Can't request without a valid account")
    }

    if (!recipient.address) {
      throw new Error('Error passing through the requestee address')
    }

    if (!usdAmount) {
      // Should not happen
      throw new Error('Trying to request from a token without a usd price')
    }

    const paymentInfo = {
      amount: usdAmount.toString(),
      comment: comment || undefined,
      createdAt: firebase.database.ServerValue.TIMESTAMP,
      requesterAddress: requesterAddress,
      requesterE164Number: requesterE164Number ?? undefined,
      requesteeAddress: recipient.address.toLowerCase(),
      status: PaymentRequestStatus.REQUESTED,
      notified: false,
    }

    ValoraAnalytics.track(RequestEvents.request_confirm_request, {
      requesteeAddress: paymentInfo.requesteeAddress,
    })
    dispatch(writePaymentRequest(paymentInfo))
    Logger.showMessage(t('requestSent'))
  }

  const renderFooter = () => {
    return (
      <View style={styles.feeContainer}>
        <TokenTotalLineItem tokenAmount={tokenAmount} tokenAddress={transactionData.tokenAddress} />
      </View>
    )
  }

  return (
    <SafeAreaView style={styles.container} edges={['bottom']}>
      <DisconnectBanner />
      <ReviewFrame
        FooterComponent={renderFooter}
        confirmButton={{
          action: onConfirm,
          text: t('request'),
          disabled: false,
        }}
      >
        <View style={styles.transferContainer}>
          <View style={styles.headerContainer}>
            <ContactCircle recipient={recipient} />
            <View style={styles.recipientInfoContainer}>
              <Text style={styles.headerText}>{t('requesting')}</Text>
              <Text style={styles.displayName}>{getDisplayName(recipient, t)}</Text>
            </View>
          </View>
          <TokenDisplay
            style={styles.amount}
            amount={tokenAmount}
            tokenAddress={transactionData.tokenAddress}
          />
          <CommentTextInput
            testID={'request'}
            onCommentChange={setComment}
            comment={comment}
            onBlur={onBlur}
          />
        </View>
      </ReviewFrame>
    </SafeAreaView>
  )
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    padding: 8,
  },
  feeContainer: {
    padding: 16,
    paddingBottom: 8,
  },
  transferContainer: {
    alignItems: 'flex-start',
    paddingBottom: 24,
  },
  headerContainer: {
    flexDirection: 'row',
    alignItems: 'flex-start',
  },
  recipientInfoContainer: {
    paddingLeft: 8,
  },
  headerText: {
    ...fontStyles.regular,
    color: colors.gray4,
  },
  displayName: {
    ...fontStyles.regular500,
  },
  amount: {
    paddingVertical: 8,
    ...fontStyles.largeNumber,
  },
})

export default PaymentRequestConfirmation
