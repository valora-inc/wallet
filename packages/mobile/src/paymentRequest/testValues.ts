import { PaymentRequest, PaymentRequestStatus } from 'src/paymentRequest/types'

export function createMockPaymentRequest(partial: object): PaymentRequest {
  return {
    uid: '7363bvby2ba8270273',
    amount: '20',
    comment: 'Just the best',
    requesteeAddress: '0x15280126303735b625',
    createdAt: Date.now(),
    requesterAddress: '101929292929',
    status: PaymentRequestStatus.REQUESTED,
    notified: true,
    ...partial,
  }
}
