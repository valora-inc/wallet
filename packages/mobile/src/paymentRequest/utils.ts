import { hexToBuffer } from '@celo/utils/lib/address'
import { decryptComment, encryptComment } from '@celo/utils/lib/commentEncryption'
import { isE164Number } from '@celo/utils/lib/phoneNumbers'
import { call } from 'redux-saga/effects'
import { MAX_COMMENT_LENGTH } from 'src/config'
import { features } from 'src/flags'
import i18n from 'src/i18n'
import { PaymentRequest, WriteablePaymentRequest } from 'src/paymentRequest/types'
import Logger from 'src/utils/Logger'
import { doFetchDataEncryptionKey } from 'src/web3/dataEncryptionKey'

const TAG = 'paymentRequest/utils'

// Encrypt sensitive data in the payment request using the recipient and sender DEK
export function* encryptPaymentRequest(paymentRequest: WriteablePaymentRequest) {
  Logger.debug(`${TAG}@encryptPaymentRequest`, 'Encrypting payment request')

  const fromKey: Buffer | null = yield call(
    doFetchDataEncryptionKey,
    paymentRequest.requesterAddress
  )
  if (!fromKey) {
    Logger.debug(`${TAG}@encryptPaymentRequest`, 'No sender key found, skipping encryption')
    return sanitizePaymentRequest(paymentRequest)
  }

  const toKey: Buffer | null = yield call(doFetchDataEncryptionKey, paymentRequest.requesteeAddress)
  if (!toKey) {
    Logger.debug(`${TAG}@encryptPaymentRequest`, 'No recipient key found, skipping encryption')
    return sanitizePaymentRequest(paymentRequest)
  }

  const encryptedPaymentRequest: WriteablePaymentRequest = {
    ...paymentRequest,
  }

  if (paymentRequest.requesterE164Number) {
    // Using the same util as for comment encryption to encrypt the phone number
    // TODO: Consider renaming this util for clarity
    const { comment: encryptedE164Number, success } = encryptComment(
      paymentRequest.requesterE164Number,
      toKey,
      fromKey
    )

    // We intentionally exclude the phone number if we can't encrypt it
    // The request still contains an address which may still map to a contact
    // If the recipient has seen it before.
    encryptedPaymentRequest.requesterE164Number = success ? encryptedE164Number : undefined
  }

  const comment = paymentRequest.comment
  if (comment && features.USE_COMMENT_ENCRYPTION) {
    const { comment: encryptedComment, success } = encryptComment(comment, toKey, fromKey)
    encryptedPaymentRequest.comment = success ? encryptedComment : comment
  }

  return encryptedPaymentRequest
}

// Decrypt sensitive data in the payment request using the user's DEK
export function decryptPaymentRequest(
  paymentRequest: PaymentRequest,
  dataEncryptionKey: string | null,
  isOutgoingRequest: boolean
) {
  Logger.debug(`${TAG}@decryptPaymentRequest`, 'Decrypting payment request')

  if (!dataEncryptionKey) {
    Logger.error(`${TAG}@decryptPaymentRequest`, 'Missing DEK, should never happen.')
    return paymentRequest
  }
  const dekBuffer = hexToBuffer(dataEncryptionKey)

  const decryptedPaymentRequest: PaymentRequest = {
    ...paymentRequest,
  }

  const requesterE164Number = paymentRequest.requesterE164Number
  if (requesterE164Number) {
    const { comment: decryptedRequesterE164Number, success } = decryptComment(
      requesterE164Number,
      dekBuffer,
      isOutgoingRequest
    )
    if (success) {
      decryptedPaymentRequest.requesterE164Number = decryptedRequesterE164Number
    } else if (isE164Number(requesterE164Number)) {
      Logger.warn(
        `${TAG}@decryptPaymentRequest`,
        'Decrypting requesterE164Number failed, using raw number'
      )
      decryptedPaymentRequest.requesterE164Number = requesterE164Number
    } else {
      Logger.warn(
        `${TAG}@decryptPaymentRequest`,
        'requesterE164Number appears to be ciphertext, excluding it'
      )
      decryptedPaymentRequest.requesterE164Number = undefined
    }
  }

  const comment = paymentRequest.comment
  if (comment && features.USE_COMMENT_ENCRYPTION) {
    const { comment: decryptedComment, success } = decryptComment(
      comment,
      dekBuffer,
      isOutgoingRequest
    )
    if (success) {
      decryptedPaymentRequest.comment = decryptedComment
    } else if (comment.length <= MAX_COMMENT_LENGTH) {
      Logger.warn(`${TAG}@decryptPaymentRequest`, 'Decrypting comment failed, using raw comment')
      decryptedPaymentRequest.comment = comment
    } else {
      Logger.warn(
        `${TAG}@decryptPaymentRequest`,
        'Comment appears to be ciphertext, hiding comment'
      )
      decryptedPaymentRequest.comment = i18n.t('commentUnavailable')
    }
  }

  return decryptedPaymentRequest
}

// For cases when the request can't be encrypted, remove sensitive PII
function sanitizePaymentRequest(paymentRequest: WriteablePaymentRequest): WriteablePaymentRequest {
  return {
    ...paymentRequest,
    requesterE164Number: undefined,
  }
}
