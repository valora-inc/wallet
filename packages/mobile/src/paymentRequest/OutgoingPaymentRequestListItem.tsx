import * as React from 'react'
import { WithTranslation } from 'react-i18next'
import { StyleSheet, View } from 'react-native'
import { HomeEvents } from 'src/analytics/Events'
import ValoraAnalytics from 'src/analytics/ValoraAnalytics'
import ContactCircle from 'src/components/ContactCircle'
import CurrencyDisplay from 'src/components/CurrencyDisplay'
import RequestMessagingCard from 'src/components/RequestMessagingCard'
import { NotificationBannerCTATypes, NotificationBannerTypes } from 'src/home/NotificationBox'
import { withTranslation } from 'src/i18n'
import { cancelPaymentRequest, updatePaymentRequestNotified } from 'src/paymentRequest/actions'
import { getDisplayName, Recipient } from 'src/recipients/recipient'
import { Currency } from 'src/utils/currencies'
import Logger from 'src/utils/Logger'

interface OwnProps {
  id: string
  requestee: Recipient
  amount: string
  comment?: string
  cancelPaymentRequest: typeof cancelPaymentRequest
  updatePaymentRequestNotified: typeof updatePaymentRequestNotified
}

type Props = OwnProps & WithTranslation

export class OutgoingPaymentRequestListItem extends React.Component<Props> {
  onRemind = () => {
    const { id, t } = this.props
    this.props.updatePaymentRequestNotified(id, false)
    ValoraAnalytics.track(HomeEvents.notification_select, {
      notificationType: NotificationBannerTypes.outgoing_tx_request,
      selectedAction: NotificationBannerCTATypes.remind,
    })
    Logger.showMessage(t('reminderSent'))
  }

  onCancel = () => {
    const { id } = this.props
    ValoraAnalytics.track(HomeEvents.notification_select, {
      notificationType: NotificationBannerTypes.outgoing_tx_request,
      selectedAction: NotificationBannerCTATypes.decline,
    })
    this.props.cancelPaymentRequest(id)
  }

  getCTA = () => {
    return [
      {
        text: this.props.t('remind'),
        onPress: this.onRemind,
      },
      {
        text: this.props.t('cancel'),
        onPress: this.onCancel,
      },
    ]
  }

  render() {
    const { requestee, id, comment, t } = this.props
    const amount = {
      value: this.props.amount,
      currencyCode: Currency.Dollar,
    }

    return (
      <View style={styles.container}>
        <RequestMessagingCard
          testID={`OutgoingPaymentRequestNotification/${id}`}
          title={t('outgoingPaymentRequestNotificationTitle', {
            name: getDisplayName(requestee, t),
          })}
          amount={<CurrencyDisplay amount={amount} />}
          details={comment}
          icon={<ContactCircle recipient={requestee} />}
          callToActions={this.getCTA()}
        />
      </View>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    marginBottom: 16,
  },
})

export default withTranslation<Props>()(OutgoingPaymentRequestListItem)
