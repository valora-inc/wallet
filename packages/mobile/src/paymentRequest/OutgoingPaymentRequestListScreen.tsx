import React from 'react'
import { WithTranslation } from 'react-i18next'
import { View } from 'react-native'
import { connect } from 'react-redux'
import i18n, { withTranslation } from 'src/i18n'
import { E164NumberToAddressType } from 'src/identity/reducer'
import { e164NumberToAddressSelector } from 'src/identity/selectors'
import {
  NotificationList,
  titleWithBalanceNavigationOptions,
} from 'src/notifications/NotificationList'
import { cancelPaymentRequest, updatePaymentRequestNotified } from 'src/paymentRequest/actions'
import OutgoingPaymentRequestListItem from 'src/paymentRequest/OutgoingPaymentRequestListItem'
import { getOutgoingPaymentRequests } from 'src/paymentRequest/selectors'
import { PaymentRequest } from 'src/paymentRequest/types'
import { getRecipientFromAddress, RecipientInfo } from 'src/recipients/recipient'
import { recipientInfoSelector } from 'src/recipients/reducer'
import { RootState } from 'src/redux/reducers'

interface StateProps {
  paymentRequests: PaymentRequest[]
  e164PhoneNumberAddressMapping: E164NumberToAddressType
  recipientInfo: RecipientInfo
}

interface DispatchProps {
  cancelPaymentRequest: typeof cancelPaymentRequest
  updatePaymentRequestNotified: typeof updatePaymentRequestNotified
}

const mapStateToProps = (state: RootState): StateProps => ({
  paymentRequests: getOutgoingPaymentRequests(state),
  e164PhoneNumberAddressMapping: e164NumberToAddressSelector(state),
  recipientInfo: recipientInfoSelector(state),
})

type Props = WithTranslation & StateProps & DispatchProps

export const listItemRenderer = (params: {
  recipientInfo: RecipientInfo
  cancelPaymentRequest: typeof cancelPaymentRequest
  updatePaymentRequestNotified: typeof updatePaymentRequestNotified
}) => (request: PaymentRequest, key: number | undefined = undefined) => {
  const requestee = getRecipientFromAddress(request.requesteeAddress, params.recipientInfo)
  return (
    <View key={key}>
      <OutgoingPaymentRequestListItem
        id={request.uid || ''}
        amount={request.amount}
        requestee={requestee}
        comment={request.comment}
        cancelPaymentRequest={params.cancelPaymentRequest}
        updatePaymentRequestNotified={params.updatePaymentRequestNotified}
      />
    </View>
  )
}

const OutgoingPaymentRequestListScreen = (props: Props) => {
  return (
    <NotificationList items={props.paymentRequests} listItemRenderer={listItemRenderer(props)} />
  )
}

OutgoingPaymentRequestListScreen.navigationOptions = titleWithBalanceNavigationOptions(
  i18n.t('outgoingPaymentRequests')
)

export default connect<StateProps, DispatchProps, {}, RootState>(mapStateToProps, {
  cancelPaymentRequest,
  updatePaymentRequestNotified,
})(withTranslation<Props>()(OutgoingPaymentRequestListScreen))
