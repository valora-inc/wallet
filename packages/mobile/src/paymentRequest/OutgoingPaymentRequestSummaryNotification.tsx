import * as React from 'react'
import { WithTranslation } from 'react-i18next'
import { Image } from 'react-native'
import { connect } from 'react-redux'
import { HomeEvents } from 'src/analytics/Events'
import ValoraAnalytics from 'src/analytics/ValoraAnalytics'
import { NotificationBannerCTATypes, NotificationBannerTypes } from 'src/home/NotificationBox'
import { withTranslation } from 'src/i18n'
import { E164NumberToAddressType } from 'src/identity/reducer'
import { e164NumberToAddressSelector } from 'src/identity/selectors'
import { notificationOutgoingRequest } from 'src/images/Images'
import { navigate } from 'src/navigator/NavigationService'
import { Screens } from 'src/navigator/Screens'
import SummaryNotification from 'src/notifications/SummaryNotification'
import { cancelPaymentRequest, updatePaymentRequestNotified } from 'src/paymentRequest/actions'
import { listItemRenderer } from 'src/paymentRequest/OutgoingPaymentRequestListScreen'
import PaymentRequestNotificationInner from 'src/paymentRequest/PaymentRequestNotificationInner'
import { PaymentRequest } from 'src/paymentRequest/types'
import { getRecipientFromAddress, RecipientInfo } from 'src/recipients/recipient'
import { recipientInfoSelector } from 'src/recipients/reducer'
import { RootState } from 'src/redux/reducers'

interface OwnProps {
  requests: PaymentRequest[]
}

interface DispatchProps {
  cancelPaymentRequest: typeof cancelPaymentRequest
  updatePaymentRequestNotified: typeof updatePaymentRequestNotified
}

interface StateProps {
  e164PhoneNumberAddressMapping: E164NumberToAddressType
  recipientInfo: RecipientInfo
}

type Props = OwnProps & DispatchProps & WithTranslation & StateProps

const mapStateToProps = (state: RootState): StateProps => ({
  e164PhoneNumberAddressMapping: e164NumberToAddressSelector(state),
  recipientInfo: recipientInfoSelector(state),
})

// Payment Request notification for the notification center on home screen
export class OutgoingPaymentRequestSummaryNotification extends React.Component<Props> {
  onReview = () => {
    ValoraAnalytics.track(HomeEvents.notification_select, {
      notificationType: NotificationBannerTypes.outgoing_tx_request,
      selectedAction: NotificationBannerCTATypes.review,
    })
    navigate(Screens.OutgoingPaymentRequestListScreen)
  }

  itemRenderer = (item: PaymentRequest) => {
    return (
      <PaymentRequestNotificationInner
        key={item.uid}
        amount={item.amount}
        recipient={getRecipientFromAddress(item.requesteeAddress, this.props.recipientInfo)}
        t={this.props.t}
      />
    )
  }

  render() {
    const { requests, t } = this.props
    return requests.length === 1 ? (
      listItemRenderer({
        recipientInfo: this.props.recipientInfo,
        // accessing via this.props.<...> to avoid shadowing
        cancelPaymentRequest: this.props.cancelPaymentRequest,
        updatePaymentRequestNotified: this.props.updatePaymentRequestNotified,
      })(requests[0])
    ) : (
      <SummaryNotification<PaymentRequest>
        items={requests}
        title={t('outgoingPaymentRequestsSummaryTitle', { count: requests.length })}
        detailsI18nKey="outgoingPaymentRequestsSummaryDetails"
        icon={<Image source={notificationOutgoingRequest} resizeMode="contain" />}
        onReview={this.onReview}
        itemRenderer={this.itemRenderer}
      />
    )
  }
}

export default connect<StateProps, DispatchProps, {}, RootState>(mapStateToProps, {
  cancelPaymentRequest,
  updatePaymentRequestNotified,
})(withTranslation<Props>()(OutgoingPaymentRequestSummaryNotification))
