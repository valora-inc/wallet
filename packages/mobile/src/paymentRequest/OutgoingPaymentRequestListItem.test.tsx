import { render } from '@testing-library/react-native'
import { noop } from 'lodash'
import * as React from 'react'
import 'react-native'
import { Provider } from 'react-redux'
import { cancelPaymentRequest, updatePaymentRequestNotified } from 'src/paymentRequest/actions'
import OutgoingPaymentRequestListItem from 'src/paymentRequest/OutgoingPaymentRequestListItem'
import { createMockStore } from 'test/utils'
const store = createMockStore()

const commonProps = {
  id: '1',
  amount: '24',
  comment: 'Hey thanks for the loan, Ill pay you back ASAP. LOVE YOU',
  requestee: {
    e164PhoneNumber: '5126608970',
    displayId: '5126608970',
    address: '0x91623f625e23ac1400',
    name: '5126608970',
    contact: undefined,
  },
  cancelPaymentRequest: noop as typeof cancelPaymentRequest,
  updatePaymentRequestNotified: noop as typeof updatePaymentRequestNotified,
}

describe('OutgoingPaymentRequestListItem', () => {
  it('renders correctly', () => {
    const tree = render(
      <Provider store={store}>
        <OutgoingPaymentRequestListItem {...commonProps} />
      </Provider>
    )

    expect(tree).toMatchSnapshot()
  })
})
