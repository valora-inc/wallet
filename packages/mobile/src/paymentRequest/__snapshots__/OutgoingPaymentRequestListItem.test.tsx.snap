// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OutgoingPaymentRequestListItem renders correctly 1`] = `
<View
  style={
    Object {
      "marginBottom": 16,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#FFFFFF",
          "padding": 16,
        },
        Object {
          "borderRadius": 8,
        },
        Object {
          "elevation": 12,
          "shadowColor": "rgba(156, 164, 169, 0.4)",
          "shadowOffset": Object {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 1,
          "shadowRadius": 12,
        },
        Array [
          Object {
            "flex": 1,
            "minHeight": 144,
          },
          Object {},
        ],
      ]
    }
    testID="OutgoingPaymentRequestNotification/1"
  >
    <View
      style={
        Object {
          "flex": 1,
          "flexDirection": "row",
        }
      }
    >
      <View
        style={
          Object {
            "flex": 1,
            "marginRight": 16,
          }
        }
      >
        <Text
          numberOfLines={2}
          style={
            Object {
              "color": "#9CA4A9",
              "fontFamily": "Inter-Regular",
              "fontSize": 13,
              "lineHeight": 16,
              "marginBottom": 4,
            }
          }
          testID="OutgoingPaymentRequestNotification/1/Title"
        >
          outgoingPaymentRequestNotificationTitle, {"name":"5126608970"}
        </Text>
        <Text
          numberOfLines={1}
          style={
            Object {
              "color": "#2E3338",
              "fontFamily": "Inter-SemiBold",
              "fontSize": 24,
              "lineHeight": 28,
              "marginBottom": 2,
            }
          }
          testID="OutgoingPaymentRequestNotification/1/Amount"
        >
          <Text
            style={
              Array [
                undefined,
                Object {
                  "color": undefined,
                },
              ]
            }
            testID="undefined/value"
          >
            
            â‚±
            31.92
          </Text>
        </Text>
        <Text
          style={
            Object {
              "color": "#2E3338",
              "fontFamily": "Inter-Regular",
              "fontSize": 14,
              "lineHeight": 18,
            }
          }
          testID="OutgoingPaymentRequestNotification/1/Details"
        >
          Hey thanks for the loan, Ill pay you back ASAP. LOVE YOU
        </Text>
      </View>
      <View
        style={Object {}}
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "flexDirection": "row",
                "justifyContent": "center",
              },
              undefined,
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "justifyContent": "center",
                },
                Object {
                  "backgroundColor": "hsl(166, 53%, 93%)",
                  "borderRadius": 20,
                  "height": 40,
                  "width": 40,
                },
              ]
            }
          >
            <Text
              allowFontScaling={false}
              style={
                Array [
                  Object {
                    "color": "#FFFFFF",
                    "fontFamily": "Inter-Medium",
                    "fontSize": 16,
                  },
                  Object {
                    "color": "hsl(166, 67%, 24%)",
                    "fontSize": 20,
                  },
                ]
              }
            >
              5
            </Text>
          </View>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "flexDirection": "row",
          "flexWrap": "wrap",
          "justifyContent": "flex-start",
        }
      }
      testID="OutgoingPaymentRequestNotification/1/CallToActions"
    >
      <View
        accessible={true}
        focusable={true}
        nativeBackgroundAndroid={
          Object {
            "attribute": "selectableItemBackgroundBorderless",
            "rippleRadius": undefined,
            "type": "ThemeAttrAndroid",
          }
        }
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        testID="OutgoingPaymentRequestNotification/1/CallToActions/remind/Button"
      >
        <Text
          style={
            Array [
              Object {
                "color": "#1AB775",
                "fontFamily": "Inter-SemiBold",
                "fontSize": 16,
                "lineHeight": 22,
              },
              Object {
                "fontSize": 14,
                "lineHeight": 16,
                "marginRight": 24,
                "minHeight": 16,
                "minWidth": 48,
              },
            ]
          }
        >
          remind
        </Text>
      </View>
      <View
        accessible={true}
        focusable={true}
        nativeBackgroundAndroid={
          Object {
            "attribute": "selectableItemBackgroundBorderless",
            "rippleRadius": undefined,
            "type": "ThemeAttrAndroid",
          }
        }
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        testID="OutgoingPaymentRequestNotification/1/CallToActions/cancel/Button"
      >
        <Text
          style={
            Array [
              Object {
                "color": "#1AB775",
                "fontFamily": "Inter-SemiBold",
                "fontSize": 16,
                "lineHeight": 22,
              },
              Object {
                "fontSize": 14,
                "lineHeight": 16,
                "marginRight": 24,
                "minHeight": 16,
                "minWidth": 48,
              },
            ]
          }
        >
          cancel
        </Text>
      </View>
    </View>
  </View>
</View>
`;
